test/controllers/BankDetailsControllerSpec.scala

```diff
@@ -0,0 +1,249 @@
+          ("sortCode", "123456"),
+          ("accountNumber", "12345678"),
+          ("isSoleSignatory", "Yes")
+        )
+      val result: Future[Result] = controller.enterBankDetailsSubmit(fakeRequest)
+      status(result) shouldBe Status.SEE_OTHER
+      redirectLocation(result) shouldBe Some(PageUrls.checkDirectDebitDetailsUrl)
+      EssttpBackend.DirectDebitDetails.verifyUpdateDirectDebitDetailsRequest(TdAll.journeyId)
+    }
+
+    "redirect to /you-cannot-set-up-a-direct-debit-online when user submits no for radio option relating to being account holder" in {
+      AuthStub.authorise()
+      EssttpBackend.HasCheckedPlan.findJourneyAfterUpdateHasCheckedPlan()
+      EssttpBackend.DirectDebitDetails.updateDirectDebitDetails(TdAll.journeyId)
+      val fakeRequest = FakeRequest(
+        method = "POST",
+        path   = "/set-up-direct-debit"
+      ).withAuthToken()
+        .withSession(SessionKeys.sessionId -> "IamATestSessionId")
+        .withFormUrlEncodedBody(
+          ("name", "Bob Ross"),
+          ("sortCode", "123456"),
+          ("accountNumber", "12345678"),
+          ("isSoleSignatory", "No")
+        )
+      val result: Future[Result] = controller.enterBankDetailsSubmit(fakeRequest)
+      status(result) shouldBe Status.SEE_OTHER
+      redirectLocation(result) shouldBe Some(PageUrls.cannotSetupDirectDebitOnlineUrl)
+      EssttpBackend.DirectDebitDetails.verifyUpdateDirectDebitDetailsRequest(TdAll.journeyId)
+    }
+
+    "show correct error messages when form submitted is empty" in {
+      AuthStub.authorise()
+      EssttpBackend.HasCheckedPlan.findJourneyAfterUpdateHasCheckedPlan()
+
+      val fakeRequest = FakeRequest(
+        method = "POST",
+        path   = "/set-up-direct-debit"
+      ).withAuthToken()
+        .withSession(SessionKeys.sessionId -> "IamATestSessionId")
+        .withFormUrlEncodedBody(
+          ("name", ""),
+          ("sortCode", ""),
+          ("accountNumber", ""),
+          ("isSoleSignatory", "")
+        )
+      val result: Future[Result] = controller.enterBankDetailsSubmit(fakeRequest)
+      status(result) shouldBe Status.OK
+      EssttpBackend.DirectDebitDetails.verifyNoneUpdateDirectDebitDetailsRequest(TdAll.journeyId)
+
+      val pageContent: String = contentAsString(result)
+      val doc: Document = Jsoup.parse(pageContent)
+
+      val errorSummary = doc.select(".govuk-error-summary__list")
+      val errorLink = errorSummary.select("a").asScala.toList
+      errorLink(0).text() shouldBe "Enter the name on the account"
+      errorLink(0).attr("href") shouldBe accountNameFieldId
+      errorLink(1).text() shouldBe "Enter sort code"
+      errorLink(1).attr("href") shouldBe sortCodeFieldId
+      errorLink(2).text() shouldBe "Enter account number"
+      errorLink(2).attr("href") shouldBe accountNumberFieldId
+      errorLink(3).text() shouldBe "Select yes if you are the account holder"
+      errorLink(3).attr("href") shouldBe accountHolderRadioId
```

Comment:
there's a bit of repetition here you could probably condense down into a function like:
```scala
def testFormError(
  formInput: (String,String)*
)(
  expectedErrorTextWithHrefs: List[(String, String)]
)
```

Reply:
I agree but felt it was a bit more readable that way, happy to change it though